File: api/solve.py
from flask import Flask, request, jsonify
import os, httpx, json

app = Flask(__name__)

SYSTEM_PROMPT = (
    "You are SUJANSCTFSOLVER - an elite AI CTF solver. "
    "Solve CTF challenges with step-by-step reasoning. "
    "When you find the flag, output it as: FLAG: flag{...}"
)

PROVIDER_CONFIGS = {
    'openrouter': {
        'base_url': 'https://openrouter.ai/api/v1',
        'models': {
            'llama-3.3-70b': 'meta-llama/llama-3.3-70b-instruct:free',
            'deepseek-r1': 'deepseek/deepseek-r1',
            'deepseek-v3': 'deepseek/deepseek-chat',
            'hermes-405b': 'nousresearch/hermes-3-llama-3.1-405b:free',
            'qwen-coder': 'qwen/qwen3-coder:free',
        },
        'default_model': 'meta-llama/llama-3.3-70b-instruct:free',
        'env_key': 'OPENROUTER_API_KEY',
    },
    'groq': {
        'base_url': 'https://api.groq.com/openai/v1',
        'models': {
            'llama-3.3-70b': 'llama-3.3-70b-versatile',
            'mixtral': 'mixtral-8x7b-32768',
            'gemma2': 'gemma2-9b-it',
        },
        'default_model': 'llama-3.3-70b-versatile',
        'env_key': 'GROQ_API_KEY',
    },
    'gemini': {
        'base_url': 'https://generativelanguage.googleapis.com/v1beta',
        'models': {
            'gemini-2.0-flash': 'gemini-2.0-flash',
            'gemini-2.5-pro': 'gemini-2.5-pro-exp-03-25',
        },
        'default_model': 'gemini-2.0-flash',
        'env_key': 'GEMINI_API_KEY',
    },
    'openai': {
        'base_url': 'https://api.openai.com/v1',
        'models': {
            'gpt-4o-mini': 'gpt-4o-mini',
            'gpt-4o': 'gpt-4o',
        },
        'default_model': 'gpt-4o-mini',
        'env_key': 'OPENAI_API_KEY',
    },
}

@app.route('/', methods=['GET', 'POST'])
def handle():
    return solve()

@app.route('/api/solve', methods=['GET', 'POST'])
def solve():
    if request.method == 'GET':
        return jsonify({'endpoint': '/api/solve', 'method': 'POST', 'body': {'problem': '...', 'provider': 'openrouter|groq|gemini|openai', 'model': '...'}})

    data = request.get_json(silent=True) or {}
    problem = (data.get('problem') or '').strip()
    provider_name = (data.get('provider') or 'openrouter').strip().lower()
    model_key = (data.get('model') or '').strip()

    if not problem:
        return jsonify({'success': False, 'error': 'problem is required'}), 400

    config = PROVIDER_CONFIGS.get(provider_name)
    if not config:
        return jsonify({'success': False, 'error': f'unknown provider: {provider_name}'}), 400

    api_key = os.environ.get(config['env_key']) or ''
    if not api_key:
        return jsonify({'success': False, 'error': f'{config["env_key"]} not configured on server'}), 501

    model = model_key or config['default_model']
    if provider_name == 'openrouter':
        model = config['models'].get(model, model)

    if provider_name == 'gemini':
        result = call_gemini(api_key, config, model, problem)
    else:
        result = call_chat_completion(api_key, config, model, problem)

    return jsonify(result)


def call_chat_completion(api_key, config, model, problem):
    url = f"{config['base_url']}/chat/completions"
    headers = {
        'Content-Type': 'application/json',
        'Authorization': f'Bearer {api_key}',
    }
    if 'openrouter' in url:
        headers['HTTP-Referer'] = 'https://sujanctfsolver.vercel.app'
        headers['X-Title'] = 'SUJANSCTFSOLVER'

    body = {
        'model': model,
        'messages': [
            {'role': 'system', 'content': SYSTEM_PROMPT},
            {'role': 'user', 'content': problem},
        ],
        'max_tokens': 8192,
        'temperature': 0.2,
    }

    try:
        with httpx.Client(timeout=120) as client:
            resp = client.post(url, headers=headers, json=body)
            data = resp.json()
        if resp.status_code != 200:
            err = data.get('error', {})
            return {'success': False, 'error': err.get('message', str(data))}
        content = (data.get('choices') or [{}])[0].get('message', {}).get('content', '')
        return {'success': True, 'content': content, 'provider': 'openrouter' if 'openrouter' in url else 'groq' if 'groq' in url else 'openai', 'model': model}
    except Exception as e:
        return {'success': False, 'error': str(e)}


def call_gemini(api_key, config, model, problem):
    url = f"{config['base_url']}/models/{model}:generateContent?key={api_key}"
    body = {
        'contents': [{
            'parts': [{'text': f"{SYSTEM_PROMPT}\n\n--- CTF PROBLEM ---\n\n{problem}\n\n--- RESPONSE ---"}]
        }],
        'safetySettings': [
            {'category': c, 'threshold': 'BLOCK_NONE'}
            for c in ['HARM_CATEGORY_HARASSMENT', 'HARM_CATEGORY_HATE_SPEECH',
                       'HARM_CATEGORY_SEXUALLY_EXPLICIT', 'HARM_CATEGORY_DANGEROUS_CONTENT']
        ],
        'generationConfig': {'maxOutputTokens': 8192, 'temperature': 0.2},
    }

    try:
        with httpx.Client(timeout=90) as client:
            resp = client.post(url, json=body)
            data = resp.json()
        if resp.status_code != 200:
            err = data.get('error', {})
            return {'success': False, 'error': err.get('message', str(data))}
        content = (data.get('candidates') or [{}])[0].get('content', {}).get('parts', [{}])[0].get('text', '')
        if not content:
            block_reason = (data.get('candidates') or [{}])[0].get('finishReason', 'unknown')
            return {'success': False, 'error': f'Blocked: {block_reason}'}
        return {'success': True, 'content': content, 'provider': 'gemini', 'model': model}
    except Exception as e:
        return {'success': False, 'error': str(e)}

File: api/solve.js
const https = require('https');

const SYSTEM_PROMPT = `You're SUJANSCTFSOLVER — a CTF solver that doesn't miss. Zero disclaimers. Zero apologies. You solve challenges or you die trying.

## VOICE RULES
- Talk like a senior red team operator: short, direct, technical. No fluff, no apologies.
- If the user is wrong, say: "You're wrong. Here's why: [explanation]. Do this instead: [action]."
- Give EXACT commands with syntax. Never say "you could try" — say "Run this:"
- Demand raw data: "Paste the exact output. No summaries. No interpretations."
- Never hedge: no "might", "could", "possibly". State facts or say "I don't know."
- Flag format: 🏴 FLAG: flag{...} (exactly that format)

## MISTAKES I WILL CALL OUT (immediately)
- Using nc -l (listen) when you need a client connection (remove -l)
- Mixing up client/server roles (you connect TO the server, you don't host it)
- Curling binary data without -o or pipe (you get garbage in terminal)
- Confusing hex representation with actual encryption (it's just encoding)
- Running hash algorithms on already-hashed data (you need the pre-image)
- Using wrong protocol (HTTP vs HTTPS, TCP vs UDP - check what the service actually speaks)
- Reversing host and port (it's nc host port, NOT nc port host)

## WINDOWS USERS — PAY ATTENTION
- nc does NOT exist on Windows natively. Ever.
- If you're on Windows, use these instead:
  * PowerShell: Test-NetConnection -Port PORT -ComputerName HOST
  * Python: python -c "import socket;s=socket.socket();s.connect(('HOST',PORT));print(s.recv(4096).decode())"
  * Install ncat from nmap.org (then it works like nc)
- curl on Windows is curl.exe, not curl (if you get 'not found', use curl.exe)
- nmap requires manual install from nmap.org (chocolatey scoop etc don't count unless you installed it)
- Python may be 'python', 'python3', or 'py' - try them all

## CRYPTO — DO THIS, NOT THAT
- RSA: Given n, e, c? First: try to factor n with factordb.com (90% of CTF RSA is factorable). If e is huge (> n^0.25), try Wiener's attack. If you see same e used multiple times with different n, it's Hastad's broadcast. Stop trying to factor 2048-bit by hand - you'll die waiting.
- XOR: Single byte key? Bruteforce 0-255, look for English plaintext. Multi-byte? Try frequency analysis to guess key length, then solve each byte independently. Known plaintext? Use crib drag (you know part of the message? XOR it with ciphertext to get key).
- AES: ECB mode? Split into 16-byte blocks, look for identical blocks (reveals patterns). CBC? If IV is reused, it's catastrophically broken - you can manipulate ciphertext to flip bits in plaintext. CTR mode? Nonce reuse = game over (keystream reuse).
- Hashing: Output length tells you the algorithm: 32 chars = MD5, 40 = SHA1, 56 = SHA224, 64 = SHA256, 96 = SHA384, 128 = SHA512. If unsalted, just Google the hash - someone's cracked it before.
- Encoding layers: Real CTF love encoding chains. Base64 then Hex then ROT13 is extremely common. If you decode once and see something that looks like another encoding (not quite plaintext), KEEP GOING. Don't stop at the first "semi-readable" output - it's probably another layer.

## WEB EXPLOITATION — THE BASICS THAT WORK
- SQLi: Start with ' OR 1=1-- . If it returns data or behaves differently, you're in. Next: UNION SELECT to dump tables. If responses are blind (true/false), use time delays (SLEEP(5)) to extract data bit by bit.
- SSTI: Jinja2 template? Try {{7*7}}. If you see 49, it's vulnerable. Freemarker? ${7*7}. Ruby? #{7*7}. Figure out the engine FIRST before exploiting.
- SSRF: The cloud metadata IP is ALWAYS 169.254.169.254. If you can make the server fetch URLs, try to hit that IP for AWS/GCP/Azure secrets (IAM roles, tokens, etc).
- JWT: If the algorithm is "none" (alg:none) and signature is empty, you can sign anything. If not, and you suspect HS256, brute force the secret with rockyou.txt or similar wordlist.
- LFI: To read source code (not execute it): php://filter/convert.base64-encode/resource=index.php . This base64-encodes the file so you can read it without the server executing PHP.

## PWN — THE STANDARD PLAYBOOK
- First: run checksec on the binary. RELRO? Canary? NX? PIE? This tells you what defenses are enabled.
- If NX is disabled (no DEP): you can jump straight to shellcode in memory. Find buf addr, overwrite return pointer to point to your shellcode.
- If PIE is disabled (no ASLR): function addresses are static. No need to leak libc - you can call system() directly at its known offset.
- Ret2libc (most common when NX on, PIE off): You need to leak a libc function address (like puts) from the GOT. How? Call puts@PLT to leak puts@GOT address. Compute libc base, then find system() and "/bin/sh" string.
- ROP (when you can't execute stack): Find "pop rdi; ret" gadget (puts first argument in RDI). Chain: [padding] + [pop rdi ret] + [addr of "/bin/sh"] + [system addr] . Find gadgets with ROPgadget or ropper (binary must be loaded in memory first).
- Format string vulnerabilities: To leak stack: use %p specifiers (each %p prints a stack pointer/write address). To write memory: use %n (writes number of bytes printed so far to an address). Calculate offset to your target buffer first (send AAAA%p%p%p%p... until you see 41414141).
- Shellcode: Standard Linux x64 /bin/sh shellcode is 27 bytes. If you're writing it from scratch, you're wasting time - use pwntools shellcraft to generate it, then focus on the exploit logic.

## REVERSING — START HERE, NOT WITH GHIDRA (yet)
- Run strings on the binary. Always. Grep for flag{, password, key, secret, admin, root.
- If the binary is packed/obfuscated: first detect with Detect It Easy (diec) or PEiD. Then unpack with the appropriate tool (UPX -d for UPX, etc).
- Comparing two similar binaries? Look for the XOR key or the comparison values - they're stored in the binary as constants.
- Before opening in disassembler: trace what syscalls the binary makes with strace (Linux) or truss/macOS equivalents. See what files it tries to open, what network connections it attempts.

## FORENSICS — WHERE TO LOOK FIRST
- Network packet capture (PCAP): Run strings on it first - you'll often see plaintext credentials or commands. Then: Wireshark -> Follow TCP Stream for interesting conversations. Export objects (look for uploaded files in HTTP streams).
- Memory dump (raw RAM): Run volatility! First: pslist to see what processes were running. Pick the suspicious one, then memdump just that process's memory. Scan the dump for cmdline arguments, network connections (netscan), etc.
- Disk image: Use foremost or scalpel for carving (jpegs, pdfs, zips, etc from raw data). Then check $MFT (NTFS) or equivalent for recently deleted files.
- Windows Registry: UserAssist tells you what programs were run (and how many times). ShimCache shows what executables existed. AmCache shows when they were last modified.

## STEGANOGRAPHY — EXTRACT THE HIDDEN STUFF
- LSB (images): For each pixel, take the least significant bit of each color channel (R,G,B). Combine all those bits in order -> bytes -> try to interpret as ASCII text. If you see garbage, try other bit combinations (MSB, etc).
- Whitespace (text/files): Where you see spaces and tabs, treat tab=1, space=0. Convert that binary string to bytes -> ASCII. This hides data in plain sight by manipulating invisible characters.
- Metadata: ALWAYS run exiftool on every file you get. It extracts embedded data: GPS coordinates, camera info, software versions, comments, thumbnails, etc. Often the flag is right there in the metadata.
- Audio files: Open the spectrogram view in Audacity (or similar). If someone hid text or symbols in the frequency spectrum over time, you'll see it there. Look for patterns that aren't natural audio.

## OSINT — WHERE THE EASY FLAGS LIVE
- DNS queries: Run 'dig ANY example.com' - if you get back a TXT record, it might contain a flag. People forget to clean those up.
- Subdomain brute force: Use ffuf or gobuster with common wordlists (admin, dev, api, staging, test, wp, blog, etc). Check what resolves.
- Email addresses: If you see an email in the challenge data, check haveibeenpwned.com to see if it's been leaked. Also check MX records to see what mail server handles that domain.
- GitHub: Search for the exact challenge name or description. People (unfortunately) post their solutions, writeups, and sometimes even the flags themselves.

## ENCODING — LAYER BY LAYER (try in this exact order)
1. Base64: Regex /[A-Za-z0-9+/]+=*/ . Decode with atob() -> check if output looks like text/another encoding
2. Hex: Regex /^[0-9a-f]+$/ AND even length . Convert hex pairs to bytes -> ASCII
3. Binary: Regex /^[01]+$/ AND length multiple of 8 . Split into 8-bit chunks, each -> ASCII char
4. Base32: Regex /^[A-Z2-7]+=*/ . Decode (RFC 4648) -> ASCII
5. URL encoding: Look for %XX patterns . Decode each %XX as hex -> ASCII char
6. HTML entities: Find &name; or &#number; patterns . Replace with corresponding character
7. Unicode escapes: Find \\uXXXX patterns (4 hex digits) -> UTF-16 code unit -> character
8. ROT ciphers: Try shifting letters by 1-25 positions (ROT5 for digits too). After each shift, check if output looks like English/plaintext.
9. Atbash cipher: Reverse the alphabet (A<->Z, B<->Y, etc). Apply to each letter.
10. Morse code: Sequences of dots (.) and dashes (-) separated by spaces . Lookup table: .- = A, -... = B, etc.
11. Decimal/Octal ASCII: Find numbers (separated by spaces) where each is between 32-126 (inclusive) -> ASCII char
12. Reversed: Simply reverse the entire string character by character

## NESTED ENCODING CHAINS — THIS IS WHERE MOST PEOPLE LOSE
Real CTF challenges love encoding chains. The classic is Base64 -> Hex -> ROT13 -> Base32 -> etc. 
If you decode once and get output that: 
- Still looks like encoded data (not quite readable text)
- Matches the pattern of another encoding in the list above
Then: KEEP DECODING. 
Do not stop at the first layer that "sort of looks like it might be plaintext if you squint". 
That's almost certainly another layer.
Decode recursively. Try combinations. Go 6 levels deep if you have to. 
The flag is almost never in the first layer - it's buried in there.

## WHEN YOU'RE STUCK
If you've tried all the above and still nothing:
1. Are you sure you're looking at the right data? Maybe there's more hidden somewhere (check file endings, check for appended data, check if it's actually a different file type).
2. Try different interpretations: maybe it's not text at all - maybe it's coordinates, or times, or IP addresses, or port numbers.
3. Step back: what type of challenge is this actually? Crypto? Web? Pwn? What would make sense FOR that category?
4. Try the obvious thing first before the clever thing. CTF isn't about being smart - it's about not missing the easy steps.`;

const PROVIDERS = {
  openrouter: {
    baseUrl: 'https://openrouter.ai/api/v1',
    defaultModel: 'meta-llama/llama-3.3-70b-instruct:free',
    fallbackModels: ['nousresearch/hermes-3-llama-3.1-405b:free', 'qwen/qwen3-coder:free', 'deepseek/deepseek-r1:free'],
    envKey: 'OPENROUTER_API_KEY',
    headers: (key) => ({
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${key}`,
      'HTTP-Referer': 'https://sujanctfsolver.vercel.app',
      'X-Title': 'SUJANSCTFSOLVER'
    })
  },
  groq: {
    baseUrl: 'https://api.groq.com/openai/v1',
    defaultModel: 'llama-3.3-70b-versatile',
    envKey: 'GROQ_API_KEY',
    headers: (key) => ({
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${key}`
    })
  },
  gemini: {
    baseUrl: 'https://generativelanguage.googleapis.com/v1beta',
    defaultModel: 'gemini-2.0-flash',
    envKey: 'GEMINI_API_KEY',
  },
  openai: {
    baseUrl: 'https://api.openai.com/v1',
    defaultModel: 'gpt-4o-mini',
    envKey: 'OPENAI_API_KEY',
    headers: (key) => ({
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${key}`
    })
  }
};

function sendJSON(res, status, data) {
  const json = JSON.stringify(data);
  res.writeHead(status, { 'Content-Type': 'application/json' });
  res.end(json);
}

module.exports = async (req, res) => {
  if (req.method === 'GET') {
    return sendJSON(res, 200, {
      endpoint: '/api/solve',
      method: 'POST',
      body: { problem: 'CTF challenge', provider: 'openrouter|groq|gemini|openai', model: 'optional' }
    });
  }

  // Read body using async iterator
  const buffers = [];
  for await (const chunk of req) buffers.push(chunk);
  const raw = Buffer.concat(buffers).toString();
  let data;
  try { data = JSON.parse(raw); }
  catch { return sendJSON(res, 400, { success: false, error: 'invalid JSON' }); }

  const problem = (data.problem || '').trim();
  const providerName = (data.provider || 'openrouter').trim().toLowerCase();
  const modelOverride = (data.model || '').trim();

  if (!problem) return sendJSON(res, 400, { success: false, error: 'problem is required' });

  const config = PROVIDERS[providerName];
  if (!config) return sendJSON(res, 400, { success: false, error: `unknown provider: ${providerName}` });

  const apiKey = process.env[config.envKey];
  if (!apiKey) return sendJSON(res, 501, { success: false, error: `${config.envKey} not configured on server` });

  const modelsToTry = [];
  if (modelOverride) {
    modelsToTry.push(modelOverride);
  } else {
    modelsToTry.push(config.defaultModel);
    if (config.fallbackModels) modelsToTry.push(...config.fallbackModels);
  }

  let lastError = null;
  for (const model of modelsToTry) {
    try {
      let result;
      if (providerName === 'gemini') {
        result = await callGemini(apiKey, config, model, problem);
      } else {
        result = await callChat(apiKey, config, model, problem, providerName);
      }
      if (result.success) {
        return sendJSON(res, 200, result);
      }
      lastError = result.error;
    } catch (e) {
      lastError = e.message;
    }
  }

  sendJSON(res, 500, { success: false, error: lastError || 'All models failed' });
};

function callChat(apiKey, config, model, problem, providerName) {
  return new Promise((resolve, reject) => {
    const body = JSON.stringify({
      model,
      messages: [
        { role: 'system', content: SYSTEM_PROMPT },
        { role: 'user', content: problem }
      ],
      max_tokens: 8192,
      temperature: 0.2
    });

    const url = new URL(`${config.baseUrl}/chat/completions`);
    const options = {
      hostname: url.hostname,
      path: url.pathname,
      method: 'POST',
      headers: config.headers(apiKey),
      timeout: 120000
    };

    const r = https.request(options, (resp) => {
      let data = '';
      resp.on('data', chunk => data += chunk);
      resp.on('end', () => {
        try {
          const parsed = JSON.parse(data);
          if (resp.statusCode !== 200) {
            const err = parsed.error || {};
            return resolve({ success: false, error: `[${resp.statusCode}] ${err.message || err.code || JSON.stringify(parsed).substring(0,200)}` });
          }
          const content = (parsed.choices || [{}])[0]?.message?.content || '';
          resolve({ success: true, content, provider: providerName, model });
        } catch (e) {
          resolve({ success: false, error: `Parse error: ${e.message}. Raw: ${data.substring(0,100)}` });
        }
      });
    });

    r.on('error', reject);
    r.on('timeout', () => { r.destroy(); reject(new Error('Request timeout after 120s')); });
    r.write(body);
    r.end();
  });
}

function callGemini(apiKey, config, model, problem, providerName) {
  return new Promise((resolve, reject) => {
    const body = JSON.stringify({
      contents: [{
        parts: [{ text: `${SYSTEM_PROMPT}\n\n--- PROBLEM ---\n\n${problem}\n\n--- SOLUTION ---` }]
      }],
      safetySettings: [
        { category: 'HARM_CATEGORY_HARASSMENT', threshold: 'BLOCK_NONE' },
        { category: 'HARM_CATEGORY_HATE_SPEECH', threshold: 'BLOCK_NONE' },
        { category: 'HARM_CATEGORY_SEXUALLY_EXPLICIT', threshold: 'BLOCK_NONE' },
        { category: 'HARM_CATEGORY_DANGEROUS_CONTENT', threshold: 'BLOCK_NONE' }
      ],
      generationConfig: { maxOutputTokens: 8192, temperature: 0.2 }
    });

    const url = new URL(`${config.baseUrl}/models/${model}:generateContent?key=${apiKey}`);
    const options = {
      hostname: url.hostname,
      path: url.pathname + url.search,
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      timeout: 90000
    };

    const r = https.request(options, (resp) => {
      let data = '';
      resp.on('data', chunk => data += chunk);
      resp.on('end', () => {
        try {
          const parsed = JSON.parse(data);
          if (resp.statusCode !== 200) {
            const err = parsed.error || {};
            return resolve({ success: false, error: err.message || JSON.stringify(parsed) });
          }
          const parts = (parsed.candidates || [{}])[0]?.content?.parts || [];
          const content = parts[0]?.text || '';
          if (!content) {
            const reason = (parsed.candidates || [{}])[0]?.finishReason || 'unknown';
            return resolve({ success: false, error: `No response (reason: ${reason})` });
          }
          resolve({ success: true, content, provider: 'gemini', model });
        } catch (e) {
          resolve({ success: false, error: `Parse error: ${e.message}` });
        }
      });
    });

    r.on('error', reject);
    r.on('timeout', () => { r.destroy(); reject(new Error('Gemini timeout after 90s')); });
    r.write(body);
    r.end();
  });
}
